lrmd/lrmd.c

```diff
@@ -160,11 +163,102 @@ build_rsc_from_xml(xmlNode * msg)
     return rsc;
 }
 
+static void
+dup_attr(gpointer key, gpointer value, gpointer user_data)
+{
+    g_hash_table_replace(user_data, strdup(key), strdup(value));
+}
+
+static gboolean
+valid_version_format(const char *version)
+{
+    if (version == NULL) {
+        return FALSE;
+    }
+
+    if (version_format_regex == NULL) {
+        const char *regex_string = "^[[:digit:]]+([.][[:digit:]]+)*$";
+        version_format_regex = calloc(1, sizeof(regex_t));
+        regcomp(version_format_regex, regex_string, REG_EXTENDED);
+    }
+
+    return regexec(version_format_regex, version, 0, NULL, 0) == 0;
+}
+
+static const char*
+get_ra_version(const char *class, const char *provider, const char *type, gboolean update_hash)
+{
+    int len = 0;
+    char *key = NULL;
+    const char *default_version = "0.1";
+    const char *version = NULL;
+
+    CRM_CHECK(type != NULL, return default_version);
+    CRM_CHECK(class != NULL, return default_version);
+
+    if (!provider) {
+        provider = "heartbeat";
+    }
```

Comment:
I never noticed that crmd does this when building its metadata cache. It's questionable to do this at all, and certainly it makes no sense when class is not "ocf" -- but I suppose for the purposes of building a hashtable key, it doesn't cause any trouble. I'm curious now whether we default class/provider in different ways in different places; that probably needs some investigation one day.

This is fine for now. In the longer term, I'd like to maybe abstract a function for "resolving" a partial resource specifier, and use it everywhere that is needed. You don't have to worry about it; I'm just giving some context.